backend/infrahub/graphql/mutations/proposed_change.py

```diff
@@ -222,12 +222,57 @@ async def mutate(
-        state.validate_review()
+        state.validate_reviewable()
 
         created_by = await proposed_change.created_by.get_peer(db=graphql_context.db)
         if created_by and created_by.id == graphql_context.active_account_session.account_id:
             raise ValidationError(input_value="You cannot review your own proposed changes")
 
+        current_user = await NodeManager.get_one_by_id_or_default_filter(
+            id=graphql_context.active_account_session.account_id,
+            kind=InfrahubKind.GENERICACCOUNT,
+            db=graphql_context.db,
+        )
+
+        approved_by = await proposed_change.approved_by.get_peers(db=graphql_context.db)
+        rejected_by = await proposed_change.rejected_by.get_peers(db=graphql_context.db)
+        approved_by_ids = [node.id for _, node in approved_by.items()]
+        rejected_by_ids = [node.id for _, node in rejected_by.items()]
+
+        async with graphql_context.db.start_session() as db:
+            if data.decision == proposed_change_constants.ProposedChangeApprovalDecision.APPROVE:
+                if current_user.id in approved_by_ids:
+                    raise ValidationError(input_value="You have already approved this proposed change")
+                await proposed_change.approved_by.add(db=db, data=current_user)
+                if current_user.id in rejected_by_ids:
+                    await proposed_change.rejected_by.remove_locally(db=db, peer_id=current_user.id)
+
+            elif data.decision == proposed_change_constants.ProposedChangeApprovalDecision.UNDO_APPROVE:
+                if current_user.id not in approved_by_ids:
+                    raise ValidationError(
+                        input_value="You did not approve this proposed change yet, it can't be un-approved"
+                    )
+                await proposed_change.approved_by.remove_locally(db=db, peer_id=current_user.id)
+
+            elif data.decision == proposed_change_constants.ProposedChangeApprovalDecision.REJECT:
+                if current_user.id in rejected_by_ids:
+                    raise ValidationError(input_value="You have already rejected this proposed change")
+                await proposed_change.rejected_by.add(db=db, data=current_user)
+                if current_user.id in approved_by_ids:
+                    await proposed_change.approved_by.remove_locally(db=db, peer_id=current_user.id)
+
+            elif data.decision == proposed_change_constants.ProposedChangeApprovalDecision.UNDO_REJECT:
+                if current_user.id not in rejected_by_ids:
+                    raise ValidationError(
+                        input_value="You did not reject this proposed change yet, it can't be un-rejected"
+                    )
+                await proposed_change.rejected_by.remove_locally(db=db, peer_id=current_user.id)
+
+            else:
+                raise ValidationError(input_value=f"Invalid decision {data.decision}")
```

Comment:
I suppose it's a matter of taste, but could this block be more easy to read by using a `match... case` instead?